### PR TITLE
Fix 33971: CSV uploads aren't possible when native query editing is disabled

### LINKED ISSUE #33971: CSV uploads aren't possible when native query editing is disabled
## Context
In the docs we state that [Native query editing isn’t required for uploading](https://www.metabase.com/docs/latest/databases/uploads#add-people-to-a-group-with-unrestricted-data-access-to-the-upload-schema) but if you look at the image here you can see that if Native query editing is disabled, the upload option disappears. 

Image

## Expected Behavior
The ability to upload a CSV should be independent of "Native query editing" permissions. [More context in Slack](https://metaboat.slack.com/archives/C01MS7DQKR6/p1695037878188779).

### PR DESCRIPTION
**Pull request by calherries**
Fixes https://github.com/metabase/metabase/issues/33971

FE changes: 
- Adds `canUpload()` to the database class
- uses this as the check for whether a user has write access to the upload target schema
- Adds an E2E test for this specific scenario

BE changes: 
- Adds a `can_upload` boolean field to database objects for GET /api/database and GET /api/database/:id responses. Reuses the logic from `POST api/card/from-csv` for checking uploads are possible, which checks things like schema permissions and whether the uploads schema is syncable.
- Adds a test for `GET /api/database/:id` and `GET /api/database` endpoints to check the `can_upload` is true/false depending on the user's schema permissions

### CODE REVIEW

File: test/metabase/api/card_test.clj [hunk 1]
```diff
@@ -2873,19 +2873,20 @@
    (upload-example-csv! collection-id true))
   ([collection-id grant-permission?]
    (mt/with-current-user (mt/user->id :rasta)
-     (let [file              (upload-test/csv-file-with
+     (let [;; Make the file-name unique so the table names don't collide
+           csv-file-name     (str (random-uuid) ".csv")
```
Comment (calherries) [inline test/metabase/api/card_test.clj:2877]: This was an issue when I updated `metabase-enterprise.advanced-permissions.common-test/upload-csv-test` to call this function multiple times in a `doseq` block

File: src/metabase/api/card.clj [hunk 2]
```diff
@@ -1155,16 +1155,50 @@ saved later when it is ready."
        :num-columns (count (first rows))
        :num-rows    (count (rest rows))})))
 
+(defn- can-upload-error
+  "Returns an ExceptionInfo object that the user can upload to the given database and schema. Throws an error if not."
```
Comment (noahmoss) [inline src/metabase/api/card.clj:1159]: This docstring seems incorrect?

**Review by noahmoss — APPROVED** [review]
BE lgtm

**Review by alxnddr — APPROVED** [review]

File: src/metabase/api/card.clj (hunk 2 above)
Comment (calherries) [reply src/metabase/api/card.clj]: Fixed [Fix docstring](https://github.com/metabase/metabase/pull/33990/commits/7d082a57b88f2516346dd6c6164e1f4b82fe5302)

**Pull request merged**